### PR TITLE
Convert utils to ts

### PR DESCRIPTION
**Pull request by klippx**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
-let _process, getNanoSeconds, loadTime
+type HrTimeFn = () => [number, number]
+type _Process = { hrtime: HrTimeFn }
+let _process: _Process, getNanoSeconds: (() => number) | undefined, loadTime: number | undefined
 try { _process = eval('typeof __TEST_WEB__ === "undefined" && typeof process === "object" ? process : undefined') } catch (e) {} // eslint-disable-line no-eval
 
 const hasProcessHrtime = () => {
-  return (typeof _process !== 'undefined' && _process !== null) && _process.hrtime
+  return (typeof _process !== 'undefined' && _process !== null)
```
Comment (klippx) [inline src/utils.ts:7]: && _process.hrtime !== undefined

Comment (Nevon) [reply src/utils.ts:7]: This can just be `return _process?.hrtime != null`.

**Comment by tulios** [comment]
@klippx it's good to set `allowJS=true` while we migrate (https://www.typescriptlang.org/tsconfig#allowJs) and change the build script run `tsc` first

**Comment by klippx** [comment]
> @klippx it's good to set `allowJS=true` while we migrate (https://www.typescriptlang.org/tsconfig#allowJs) and change the build script run `tsc` first

Perhaps I'm making it too hard, but... should we try to change `build:browser` and `build:node` to hook typescript into the transformations? Not sure what is a good approach here.

Or like you mentioned, make build run tsc first with allowJS=true, but then we need to think about folders a bit:
Should we have `src` as a mix of TS and JS, let TSC build that to a new folder (what name? usually `dist`?)
- Then we need to change other tooling that now reads from src to read from `dist` 
- The tooling that built into `dist` (`build:browser`) will have to build to another folder, maybe`dist/web`
- The tooling that builds into lib (`build:node`) is unchanged?

File: src/utils.ts (hunk 1 above)
Comment (klippx) [reply src/utils.ts]: True, but I have been considering that in this PR I should aim to do _minimal_ changes - to not introduce any bugs or fix too many... "features"

File: .eslintrc [hunk 2]
```diff
@@ -5,5 +5,6 @@
     "jasmine": true,
     "jest": true
   },
+  "ignorePatterns": ["src/utils.js"],
```
Comment (klippx) [inline .eslintrc:8]: remove

File: tsconfig.json [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
+    "noImplicitAny": true,
+    "declaration": true,
+    "outDir": "typings",
+    "target": "es2017"
   },
   "include": [
-    "typings",
-    "spec/typescript/*",
-    "spec/typescript/**/*"
+    "typings/**/*",
```
Comment (Yurickh) [inline tsconfig.json:11]: Since `outDir` is set to `typings`, does that mean that we'll put the generated js code there, and include it back here? Isn't this a problem? 👀

File: tsconfig.json [hunk 4]
```diff
@@ -1,11 +1,19 @@
 {
   "compilerOptions": {
+    "incremental": true,
     "strict": true,
-    "noEmit": true
+    "noImplicitAny": true,
+    "declaration": true,
+    "outDir": "typings",
+    "target": "es2017"
```
Comment (Yurickh) [inline tsconfig.json:8]: Do I understand correctly this is a breaking change, since you're dropping support for engines that don't have es2017 support? (looking at you, ie11)

File: tsconfig.json (hunk 3 above)
Comment (klippx) [reply tsconfig.json:11]: Thank you, that's a fact:
```
error TS5055: Cannot write file '/Users/mathias.klippinge/src/mappersmith/typings/utils.d.ts' because it would overwrite input file.
```

Maybe we need a `tsconfig.build.json` that will not include `typings/**/*` but will instead simple emit typings to this folder.

File: tsconfig.json (hunk 4 above)
Comment (klippx) [reply tsconfig.json:8]: I can set this to `es5` to be less dramatic (and also support IE11), but in fact we already dropped integration tests for IE11 (switched to Edge instead). According to https://caniuse.com/usage-table then IE11 and lesser versions account for 0.73% of all browsers which is not much but not exactly a little either.

One benefit of dropping it is smaller bundle sizes, es5 bundle sizes are a bit large. But maybe that day should come later rather than sooner?

Comment (Yurickh) [reply tsconfig.json:8]: I'm totally in to drop ie11, no need to sell me :p~ but it _is_ a major bump

File: built/utils.d.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+export declare const lowerCaseObjectKeys: (obj: Record<string, unknown>) => Record<string, unknown>;
+export declare const assign: {
+    <T, U>(target: T, source: U): T & U;
+    <T_1, U_1, V>(target: T_1, source1: U_1, source2: V): T_1 & U_1 & V;
+    <T_2, U_2, V_1, W>(target: T_2, source1: U_2, source2: V_1, source3: W): T_2 & U_2 & V_1 & W;
+    (target: object, ...sources: any[]): any;
+};
+export declare const isPlainObject: (value: unknown) => value is Record<string, unknown>;
+export declare const btoa: (input: string) => string;
+export {};
```
Comment (klippx) [inline built/utils.d.ts:27]: wat?

File: tsconfig.json (hunk 4 above)
Comment (klippx) [reply tsconfig.json]: decided to not drop es5 in this pr, this change is already big.
